lang/csharp/src/apache/main/Specific/ObjectCreator.cs

```diff
@@ -16,164 +16,129 @@
  * limitations under the License.
  */
 using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
-using System.Text;
 using System.Reflection;
-using System.Reflection.Emit;
 
 namespace Avro.Specific
 {
-
+    /// <summary>
+    /// Resolves and creates types associated with a schema and/or name. You should generally use
+    /// the shared <see cref="Instance"/> to take advantage caching.
+    /// </summary>
     public sealed class ObjectCreator
     {
-        private static readonly ObjectCreator instance = new ObjectCreator();
-        public static ObjectCreator Instance { get { return instance; } }
+        /// <summary>
+        /// Shareable instance of the <see cref="ObjectCreator"/>.
+        /// </summary>
+        public static ObjectCreator Instance { get; } = new ObjectCreator();
 
         /// <summary>
         /// Static generic dictionary type used for creating new dictionary instances
         /// </summary>
-        private Type GenericMapType = typeof(Dictionary<,>);
+        private readonly Type GenericMapType = typeof(Dictionary<,>);
 
         /// <summary>
         /// Static generic list type used for creating new array instances
         /// </summary>
-        private Type GenericListType = typeof(List<>);
+        private readonly Type GenericListType = typeof(List<>);
 
         /// <summary>
         /// Static generic nullable type used for creating new nullable instances
         /// </summary>
-        private Type GenericNullableType = typeof(Nullable<>);
+        private readonly Type GenericNullableType = typeof(Nullable<>);
 
+        private readonly ConcurrentDictionary<string, Type> typeCacheByName;
         private readonly Assembly execAssembly;
         private readonly Assembly entryAssembly;
         private readonly bool diffAssembly;
 
-        public delegate object CtorDelegate();
```

Comment:
I removed `CtorDelegate` and `NameCtorKey` because they are not used.

Reply:
I reverted this change since it is a breaking change to the public API. I've marked these two as obsolete for now. We will remove this in v1.10.